Update LPC DMA driver to handle large data transfers

mmahadevan108 opened the pull request:
Restructure the LPC DMA driver to manager large DMA transfers. 
Also update the DMA tests to transfer large blocks of data.

mmahadevan108 committed:
drivers: dma_mcux: Rename variables in NXP MCUX driver

Rename the variables to help understand the code flow.

Signed-off-by: Mahesh Mahadevan <mahesh.mahadevan@nxp.com>

mmahadevan108 committed:
drivers: dma_mcux: Add support for big data transfers

1. Transfers are not limited by XFERCOUNT transfer length of
   LPC DMA descriptor. Added code to handle block sizes
   greater than XFERCOUNT.
2. Use the reload_en flag to decide if we should setup
   a circular descriptor chain.
3. Improve handling of source and destination width.
4. Number of DMA descriptors are defined by a Kconfig value.
5. Changed the dma_reload function to handle transfers
   greater than XFERCOUNT.

Signed-off-by: Mahesh Mahadevan <mahesh.mahadevan@nxp.com>

mmahadevan108 committed:
drivers: mcux_i2s: Use the DMA reload flag

Use the DMA reload flag to indicate we wish to use
a circular chain of DMA descriptors to reload DMA
receive buffers.

Signed-off-by: Mahesh Mahadevan <mahesh.mahadevan@nxp.com>

mmahadevan108 committed:
tests: dma: Increase data transferred in scatter_gather test

Increase the amount of data to be transferred and the
number of blocks to transfer.

Signed-off-by: Mahesh Mahadevan <mahesh.mahadevan@nxp.com>

mmahadevan108 committed:
tests: dma: Increase data transferred in loop_transfer test

Increase the amount of data to be transferred.

Signed-off-by: Mahesh Mahadevan <mahesh.mahadevan@nxp.com>

mmahadevan108 committed:
tests: dma: Enable scatter_gather test on RT685 EVK

Enable scatter gather test on NXP RT685 EVK

Signed-off-by: Mahesh Mahadevan <mahesh.mahadevan@nxp.com>

mmahadevan108 committed:
tests: dma: Enable scatter_gather test on RT595 EVK

Enable scatter gather test on NXP RT595 EVK

Signed-off-by: Mahesh Mahadevan <mahesh.mahadevan@nxp.com>

teburd commented on drivers/dma/dma_mcux_lpc.c:
```diff
@@ -89,25 +93,189 @@ static void dma_mcux_lpc_irq_handler(const struct device *dev)
+			if ((local_block.block_size / width > NXP_LPC_DMA_MAX_XFER) ||
+			    (local_block.next_block != NULL)) {
+				/* Allocate DMA descriptors */
+				next_descriptor = k_aligned_alloc(
```
Please use a descriptor pool sizable by he application for this rather than the heap

mmahadevan108 replied:
Any negatives to allocating the descriptor pool. Why would we prefer using a Kconfig?

mmahadevan108 replied:
I have updated to use `K_HEAP_DEFINE` and also provided a `Kconfig` for the application to adjust the buffer size if needed.

teburd replied:
Ah I still think this needs to be more like an object pool

See our coding guidelines, which follow misra rules. Rule 4.12 in particular I think applies here.

mmahadevan108 replied:
I have updated the code to use an array of DMA descriptors.

mmahadevan108 replied:
@teburd, do you think this can go in before this Fridays release freeze?

teburd replied:
Absolutely, tested this and it fails on intel adsp currently, but its an easy fix. See my comment with a patch. Please apply it and I will +1. Ding me on discord if I don't see it right away.

teburd commented on tests/drivers/dma/loop_transfer/src/test_dma_loop.c:
```diff
@@ -481,9 +464,13 @@ static int test_loop_repeated_start_stop(const struct device *dma)
-		if (strncmp(tx_data, rx_data[i], sizeof(rx_data[i])) != 0) {
-			return TC_FAIL;
+		TC_PRINT("RX data Loop %d\n", i);
+		for (int j = 0; j < XFER_SIZE; j++) {
```
If memcmp is available that should be used instead

mmahadevan108 replied:
Done.

teburd reviewed (commented):
Will test this on a few things before approving

mmahadevan108 committed:
boards: nxp: Update the boards using NXP LPC DMA

Switched dma_mcux_lpc DMA driver to use K_HEAP_DEFINE

Signed-off-by: Mahesh Mahadevan <mahesh.mahadevan@nxp.com>

Commit:
tests: dma: Update LLI pool size for sg test

The scatter gather test now does a transfer list of 4 and the LLI pool
config needs to be updated to match the new test requirement.

Signed-off-by: Tom Burdick <thomas.burdick@intel.com>

teburd commented:
Please add the following patch (git apply thepatch.patch) which ensures the scatter gather test works on intel platforms to the PR. Doing do, +1 from me.

```
From dbee1b3d7b16a03230fb4b74513a347c43f9c807 Mon Sep 17 00:00:00 2001
From: Tom Burdick <thomas.burdick@intel.com>
Date: Thu, 25 May 2023 18:04:13 -0500
Subject: [PATCH] tests: dma: Update LLI pool size for sg test

The scatter gather test now does a transfer list of 4 and the LLI pool
config needs to be updated to match the new test requirement.

Signed-off-by: Tom Burdick <thomas.burdick@intel.com>
---
 tests/drivers/dma/scatter_gather/boards/intel_adsp_cavs25.conf | 1 +
 1 file changed, 1 insertion(+)
 create mode 100644 tests/drivers/dma/scatter_gather/boards/intel_adsp_cavs25.conf

diff --git a/tests/drivers/dma/scatter_gather/boards/intel_adsp_cavs25.conf b/tests/drivers/dma/scatter_gather/boards/intel_adsp_cavs25.conf
new file mode 100644
index 0000000000..b2dd49710b
--- /dev/null
+++ b/tests/drivers/dma/scatter_gather/boards/intel_adsp_cavs25.conf
@@ -0,0 +1 @@
+CONFIG_DMA_DW_LLI_POOL_SIZE=4
-- 
2.38.5
```

mmahadevan108 commented:
@teburd patch applied.
